### ISSUE TITLE
lxc-attach error

### ISSUE THREAD

**Issue by lucasfs7** [fgrehm/vagrant-lxc#308]
Hi

I'm using vagrant-lxc on Arch Linux. I followed all steps on #123 and gr I run `vagrant up` I get this error:

```
There was an error executing lxc-attach

For more information on the failure, enable detailed logging by setting
the environment variable VAGRANT_LOG to DEBUG.
```

do anyone have any idea about this error?

**Comment by tuminoid** [comment]
I'm hitting this as well by using `fgrehm/centos-6-64-lxc` from VagrantCloud/Atlas.
Container will never be ready as `lxc-attach` fail and then it enters endless `lxc-info` loop.

Gist: https://gist.github.com/tuminoid/137f1b4431e89fb36fd3

**Comment by kupferk** [comment]
I also had a similar problem with Fedora 21 as my host computer. It turned out vagrant-lxc was not able to deteermine the IP address, because of two reasons:

2. The mac address of eth0 of the container cannot be determined, so the method defined in fetch_ip_from_dnsmasq_leases.rb does not work.

1. For some reasons, it seems that obtaining an IP address via DHCP takes a little bit longer than expected, so fetch_ip_with_lxc_attach.rb fails because it effectively times out.

I could temporarily fix the problem simply by increasing the timeout and the repeat count in fetch_ip_with_lxc_attach.rb.  Maybe it would help to make this value configurable?

**Comment by johnwchadwick** [comment]
I've found using the libvirt `virbr0` adapter to work when literally everything else was failing. Don't know how much of this is necessary, but here's how I got where I am:
```
sudo pacman -Sy libvirt bridge-utils
sudo systemctl enable libvirtd.service libvirt-guests.service
sudo systemctl start libvirtd.service libvirt-guests.service
sudo virsh net-autostart default
```
Then set my `/etc/lxc/default.conf` to the following:
```
lxc.network.type = veth
lxc.network.link = virbr0
lxc.network.ipv4 = 0.0.0.0/24
```

I hope some day the networking stuff is a bit more straight forward, but for now it seems piggybacking off of the libvirt stuff is a solid solution.

**Comment by bear0330** [comment]
@johnwchadwick It works for me on CentOS 7 (of course, use yum).

**Comment by Ramblurr** [comment]
In Fedora 21 I'm also running into this issue. I've got the libvirt interface setup, however the problem persists.

```
 INFO subprocess: Starting process: ["/usr/bin/sudo", "/usr/share/vagrant/gems/gems/vagrant-lxc-1.1.0/scripts/vagrant-lxc-wrapper", "lxc-attach", "--name", "VVVlxc_default_1424875316936_46758", "--namespaces", "'NETWORK|MOUNT'", "--", "/sbin/ip", "-4", "addr", "show", "scope", "global", "eth0"]
```

What do you put in your Vagrantfile for network setup?

**Comment by hak8or** [comment]
```
INFO subprocess: Starting process: ["/usr/bin/sudo", "/usr/local/bin/vagrant-lxc-wrapper", "lxc-info", "--name", "vagrant_lxc_main_1425790244109_48867"]
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stderr: /usr/local/bin/vagrant-lxc-wrapper:34: warning: Insecure world writable dir /home/hak8or in PATH, mode 040777
DEBUG subprocess: stdout: Name:           vagrant_lxc_main_1425790244109_48867
DEBUG subprocess: stdout: State:          RUNNING
DEBUG subprocess: stdout: PID:            5952
DEBUG subprocess: stdout: CPU use:        0.93 seconds
DEBUG subprocess: stdout: BlkIO use:      160.00 KiB
DEBUG subprocess: stdout: Memory use:     3.48 MiB
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 32000
DEBUG subprocess: Exit status: 0
DEBUG fetch_ip_from_dnsmasq_leases: Attempting to load ip from dnsmasq leases (mac: )
DEBUG fetch_ip_from_dnsmasq_leases: 
DEBUG fetch_ip_from_dnsmasq_leases: Ip could not be parsed from dnsmasq leases file
 INFO subprocess: Starting process: ["/usr/bin/sudo", "/usr/local/bin/vagrant-lxc-wrapper", "lxc-info", "--name", "vagrant_lxc_main_1425790244109_48867"]
DEBUG subprocess: Selecting on IO
```

I was doing what @johnwchadwick said [here](https://github.com/fgrehm/vagrant-lxc/issues/308#issuecomment-72330777) on Arch Linux but it seems that the service doesn't start.

```
sudo systemctl status libvirtd.service libvirt-guests.service
● libvirtd.service - Virtualization daemon
   Loaded: loaded (/usr/lib/systemd/system/libvirtd.service; enabled; vendor preset: disabled)
   Active: inactive (dead) since Sun 2015-03-08 00:05:08 EST; 16min ago
     Docs: man:libvirtd(8)
           http://libvirt.org
  Process: 10399 ExecStart=/usr/bin/libvirtd $LIBVIRTD_ARGS (code=exited, status=0/SUCCESS)
 Main PID: 10399 (code=exited, status=0/SUCCESS)

Mar 08 00:05:07 angstrom libvirtd[10399]: libvirt version: 1.2.13
Mar 08 00:05:07 angstrom libvirtd[10399]: Cannot check dnsmasq binary /sbin/dnsmasq: No such file or directory
Mar 08 00:05:08 angstrom libvirtd[10399]: direct firewall backend requested, but /sbin/ebtables is not available: No such file or directory
Mar 08 00:05:08 angstrom libvirtd[10399]: out of memory
Mar 08 00:05:08 angstrom libvirtd[10399]: internal error: Failed to find path for dmidecode binary
Mar 08 00:05:08 angstrom libvirtd[10399]: invalid argument: Failed to parse group 'kvm'
Mar 08 00:05:08 angstrom libvirtd[10399]: Initialization of QEMU state driver failed: invalid argument: Failed to parse group 'kvm'
Mar 08 00:05:08 angstrom libvirtd[10399]: Driver state initialization failed

● libvirt-guests.service - Suspend Active Libvirt Guests
   Loaded: loaded (/usr/lib/systemd/system/libvirt-guests.service; enabled; vendor preset: disabled)
   Active: active (exited) since Sun 2015-03-08 00:05:07 EST; 16min ago
     Docs: man:libvirtd(8)
           http://libvirt.org
  Process: 10412 ExecStart=/usr/lib/libvirt/libvirt-guests.sh start (code=exited, status=0/SUCCESS)
 Main PID: 10412 (code=exited, status=0/SUCCESS)
   CGroup: /system.slice/libvirt-guests.service
```

**Comment by hak8or** [comment]
```
INFO subprocess: Starting process: ["/usr/bin/sudo", "/usr/local/bin/vagrant-lxc-wrapper", "lxc-info", "--name", "vagrant_lxc_main_1425790244109_48867"]
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stderr: /usr/local/bin/vagrant-lxc-wrapper:34: warning: Insecure world writable dir /home/hak8or in PATH, mode 040777
DEBUG subprocess: stdout: Name:           vagrant_lxc_main_1425790244109_48867
DEBUG subprocess: stdout: State:          RUNNING
DEBUG subprocess: stdout: PID:            5952
DEBUG subprocess: stdout: CPU use:        0.93 seconds
DEBUG subprocess: stdout: BlkIO use:      160.00 KiB
DEBUG subprocess: stdout: Memory use:     3.48 MiB
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 32000
DEBUG subprocess: Exit status: 0
DEBUG fetch_ip_from_dnsmasq_leases: Attempting to load ip from dnsmasq leases (mac: )
DEBUG fetch_ip_from_dnsmasq_leases: 
DEBUG fetch_ip_from_dnsmasq_leases: Ip could not be parsed from dnsmasq leases file
 INFO subprocess: Starting process: ["/usr/bin/sudo", "/usr/local/bin/vagrant-lxc-wrapper", "lxc-info", "--name", "vagrant_lxc_main_1425790244109_48867"]
DEBUG subprocess: Selecting on IO
```

I was doing what @johnwchadwick said [here](https://github.com/fgrehm/vagrant-lxc/issues/308#issuecomment-72330777) on Arch Linux but it seems that the service doesn't start.

```
sudo systemctl status libvirtd.service libvirt-guests.service
● libvirtd.service - Virtualization daemon
   Loaded: loaded (/usr/lib/systemd/system/libvirtd.service; enabled; vendor preset: disabled)
   Active: inactive (dead) since Sun 2015-03-08 00:05:08 EST; 16min ago
     Docs: man:libvirtd(8)
           http://libvirt.org
  Process: 10399 ExecStart=/usr/bin/libvirtd $LIBVIRTD_ARGS (code=exited, status=0/SUCCESS)
 Main PID: 10399 (code=exited, status=0/SUCCESS)

Mar 08 00:05:07 angstrom libvirtd[10399]: libvirt version: 1.2.13
Mar 08 00:05:07 angstrom libvirtd[10399]: Cannot check dnsmasq binary /sbin/dnsmasq: No such file or directory
Mar 08 00:05:08 angstrom libvirtd[10399]: direct firewall backend requested, but /sbin/ebtables is not available: No such file or directory
Mar 08 00:05:08 angstrom libvirtd[10399]: out of memory
Mar 08 00:05:08 angstrom libvirtd[10399]: internal error: Failed to find path for dmidecode binary
Mar 08 00:05:08 angstrom libvirtd[10399]: invalid argument: Failed to parse group 'kvm'
Mar 08 00:05:08 angstrom libvirtd[10399]: Initialization of QEMU state driver failed: invalid argument: Failed to parse group 'kvm'
Mar 08 00:05:08 angstrom libvirtd[10399]: Driver state initialization failed

● libvirt-guests.service - Suspend Active Libvirt Guests
   Loaded: loaded (/usr/lib/systemd/system/libvirt-guests.service; enabled; vendor preset: disabled)
   Active: active (exited) since Sun 2015-03-08 00:05:07 EST; 16min ago
     Docs: man:libvirtd(8)
           http://libvirt.org
  Process: 10412 ExecStart=/usr/lib/libvirt/libvirt-guests.sh start (code=exited, status=0/SUCCESS)
 Main PID: 10412 (code=exited, status=0/SUCCESS)
   CGroup: /system.slice/libvirt-guests.service
```

Turns out that the services won't start due to [this bug](https://bugs.archlinux.org/task/35749) so you need to edit your ```/etc/libvirt/qemu.conf``` file and comment out ```group="kvm"```.

But then you need to handle the ```... firewall backend requested, but /sbin/ebtables is not ava out of memory``` issue, which can be fixed as per [here](https://bbs.archlinux.org/viewtopic.php?id=191612). 

Alright, but now etables needs to be fixed, luckily [here](https://bbs.archlinux.org/viewtopic.php?id=98258) says just to install ebtables so doing a ```yaourt -S ebtables``` installs it.

And ... damn, need dnsmasq too, and dmidecode! Throw in a ```yaourt -S dnsmasq dmidecode``` and lets try this again.

```
~/vagrant_lxc ❯❯❯ vagrant ssh
 INFO global: Vagrant version: 1.7.2
 INFO global: Ruby version: 2.0.0
 INFO global: RubyGems version: 2.0.14
 INFO global: VAGRANT_EXECUTABLE="/opt/vagrant/bin/../embedded/gems/gems/vagrant-1.7.2/bin/vagrant"
 INFO global: VAGRANT_LOG="DEBUG"
 INFO global: VAGRANT_INSTALLER_EMBEDDED_DIR="/opt/vagrant/bin/../embedded"
....
DEBUG ssh: Checking key permissions: /home/hak8or/vagrant_lxc/.vagrant/machines/main/lxc/private_key
 INFO ssh: Invoking SSH: ["vagrant@192.168.122.13", "-p", "22", "-o", "Compression=yes", "-o", "DSAAuthentication=yes", "-o", "LogLevel=FATAL", "-o", "StrictHostKeyChecking=no", "-o", "UserKnownHostsFile=/dev/null", "-o", "IdentitiesOnly=yes", "-i", "/home/hak8or/vagrant_lxc/.vagrant/machines/main/lxc/private_key"]
Welcome to Ubuntu 14.04.1 LTS (GNU/Linux 3.18.6-1-ARCH x86_64)

 * Documentation:  https://help.ubuntu.com/
vagrant@vagrant-base-trusty-amd64:~$
```

Booyeah! Works! So, thanks @johnwchadwick for the fix and all you vagrant-lxc people for making this!

So, in short, what I did was:
- Edit your ```/etc/libvirt/qemu.conf``` file and comment out ```group="kvm"```.
- Install what's needed by doing ```yaourt -S ebtables dnsmasq dmidecode```
- Do a ```vagrant destroy```
- Restart arch, do a ```vagrant up``` and viola, it works!

**Comment by sej7278** [comment]
getting the same problem on debian jessie. running the command manually seems to show its a namespaces issue:

```
sudo /usr/local/bin/vagrant-lxc-wrapper lxc-attach --name precise-lxc_default_1436783368668_80763 --namespaces 'NETWORK|MOUNT' -- /sbin/ip -4 addr show scope global eth0
sh: 1: MOUNT: not found
```

**Comment by enricostano** [comment]
Hi there, any news about this? I'm getting the same error on Arch Linux. Ping me if you need more debug info. Thanks!

**Comment by globin** [comment]
@enricostano could you run `VAGRANT_LOG=debug vagrant up` for more debug information.

Just a note: me and @fpletz are on holiday atm and at least I have quite poor internet connectivity so don't necessarily expect us to respond very fast.

**Comment by trojkat** [comment]
Full debug log from debian sid:
https://gist.github.com/trojkat/c41009debaa515188f44

**Comment by hsoft** [comment]
Now that we use `lxc-info -iH` to get IP information and thus offload that complexity to LXC directly, maybe that the issue is fixed or at least changed in nature. Maybe it's time to revisit this?

**Comment by fgrehm** [comment]
Yeah, I don't remember what exactly `lxc-attach` is used anymore for but it is about time that we "modernize" the plugin (this has been around since pre 1.0) so I'd be fine with delegating to `lxc-info` as much as we can 😄

**Comment by fgrehm** [comment]
Hey, sorry for the silence here but this project is looking for maintainers :sweat_smile: 

As per https://github.com/fgrehm/vagrant-lxc/issues/499, I've added the `ignored` label and will close this issue. Thanks for the interest in the project and LMK if you want to step up and take ownership of this project on that other issue :wave:
